README.md
# Pre-commit Hook for Technical Standards
The repository for pre-commit hooks to check Git and GitFlow conventions and other best practices

## Installation
1. Install pre-commit from https://pre-commit.com/#install

2. Create or update a `.pre-commit-config.yaml` file at the root of repository with the following content:

```
repos:
  - repo: https://github.com/one-acre-fund/oaf-pre-commit-hooks
    rev: v1.0.0
    hooks:
      - id: oaf-tech-pre-commit-hook
```
3. Auto-update the config to the latest repos' versions by executing `pre-commit autoupdate`

4. Install with `pre-commit install`

## Expected Behaviors
1. **`Git` branching model**: to make sure branch naming convention is followed
2. **`Git` log**:  to show commit messages that do not follow conventional semantics on current branch
3. **`Pre-commit` hooks**: to check whether required pre-commit hooks are installed
4. **`Gitlint`**: to validate current commit message upon [prepare-commit-msg,commit] according to `.gitlint` config directives
## Report Issues
1. Use Slack technical channels (#dev-team-leads)
2.


release_notes.md
# Release Notes

## Release v1.3.1
Use config to throttle loggging and performance-related tweaks
## Release v1.3.0
. Add a helper function `is_hook_installed_config` that parses `.pre-commit-config.yaml` to check whether required hooks are installed
## Release v1.2.2
. A fix to retrieve the right `.gitlint config file
## Release v1.2.1
. Release for other hooks and gitlint

## Release v1.1.0
. Release with pre-commit for commit-msg
## Release v1.0.0
. Initial release with pre-commit scalffold and GitFlow


tests/oaf_tech_pre_commit_hook_test.py
from __future__ import annotations

import json
import subprocess

import pytest

from pre_commit_hooks.oaf_tech_pre_commit_hook import (
    get_commits,
    get_current_branch_name,
    is_hook_installed,
    load_config,
    validate_git_commit,
)


# Tests that the function correctly identifies the pre-commit home directory and cache directory when pre_commit_home exists. tags: [happy path]
def test_load_config_pre_commit_home_exists(self, monkeypatch):
    monkeypatch.setenv("PRE_COMMIT_HOME", "/path/to/pre-commit")
    assert load_config() == 0


# Tests that the function correctly identifies the pre-commit home directory and cache directory when pre_commit_home exists. tags: [happy path]
def test_load_config_pre_commit_home_exists(self, monkeypatch):
    monkeypatch.setenv("PRE_COMMIT_HOME", "/path/to/pre-commit")
    assert load_config() == 0


# Tests that the function correctly identifies the pre-commit home directory and cache directory when pre_commit_home does not exist but xdg_cache_home exists. tags: [happy path]
def test_load_config_xdg_cache_home_exists(self, monkeypatch):
    monkeypatch.delenv("PRE_COMMIT_HOME", raising=False)
    monkeypatch.setenv("XDG_CACHE_HOME", "/path/to/cache")
    assert load_config() == 0


# Tests that the function correctly identifies the pre-commit home directory and cache directory when neither pre_commit_home nor xdg_cache_home exist but the cache directory exists. tags: [happy path]
def test_load_config_cache_dir_exists(self, monkeypatch, tmp_path):
    monkeypatch.delenv("PRE_COMMIT_HOME", raising=False)
    monkeypatch.delenv("XDG_CACHE_HOME", raising=False)
    cache_dir = tmp_path / "pre-commit"
    cache_dir.mkdir()
    assert load_config() == 0


# Tests that the function sets default values for configuration when both the live configuration and cache directory are empty. tags: [edge case]
def test_load_config_default_config_values(self, monkeypatch):
    oaf_config = {"cache": [], "live": []}
    monkeypatch.delenv("PRE_COMMIT_HOME", raising=False)
    monkeypatch.delenv("XDG_CACHE_HOME", raising=False)
    oaf_config["cache"] = []
    oaf_config["live"] = []
    assert load_config() == 10


# Tests that the function successfully downloads and saves the configuration file to the cache directory. tags: [happy path]
def test_load_config_config_download_success(self, monkeypatch, mocker):
    monkeypatch.delenv("PRE_COMMIT_HOME", raising=False)
    monkeypatch.delenv("XDG_CACHE_HOME", raising=False)
    mocker.patch("urllib.request.urlopen")
    urlopen_mock = mocker.MagicMock()
    urlopen_mock.__enter__.return_value = urlopen_mock
    urlopen_mock.read.return_value = json.dumps(
        {
            "OAF_GIT_BRANCH_NAME_REGEX": "^((release\/[0-9]+\.[0-9]+\.[0-9])|((feature|feat|cleanup|bugfix|hotfix|fix|devops)\/[A-Z]+-[0-9]+(-[A-z0-9]+)+)|(umuganda|chore)\/[a-zA-Z0-9+\-]+)$",
            "OAF_WATCH_COMMIT_HISTORY": False,
            "OAF_GIT_BRANCH_NAME_EXCEPTION": ["develop", "main", "master"],
            "OAF_GIT_COMMIT_TYPES": [
                "feat",
                "fix",
                "style",
                "refactor",
                "docs",
                "perf",
                "test",
                "chore",
                "ci",
                "build",
            ],
            "OAF_REQUIRED_HOOKS": {
                "ggshield": {
                    "args": ["--help"],
                    "repo": "https://github.com/gitguardian/ggshield",
                },
                "gitlint": {
                    "args": ["--help"],
                    "repo": "https://github.com/jorisroovers/gitlint",
                },
            },
        }
    ).encode()
    urlopen_mock.status = 200
    urlopen_mock.reason = "OK"
    urlopen_mock.headers = {}
    urlopen_mock.url = "https://raw.githubusercontent.com/one-acre-fund/oaf-pre-commit-hooks/main/config.json"
    urlopen_mock.info.return_value.get_content_charset.return_value = "utf-8"
    urlopen_mock.info.return_value.get_content_type.return_value = "application/json"
    urlopen_mock.info.return_value.get_param.return_value = None
    urlopen_mock.info.return_value.getplist.return_value = []
    urlopen_mock.info.return_value.items.return_value = []
    urlopen_mock.info.return_value.getheaders.return_value = []
    urlopen_mock.info.return_value.get_all.return_value = []
    urlopen_mock.info.return_value.typeheader = None
    urlopen_mock.info.return_value.getencoding.return_value = None
    urlopen_mock.info.return_value.getmaintype.return_value = None
    mocker.patch("ssl._create_default_https_context")
    ssl_mock = mocker.MagicMock()
    ssl_mock._create_unverified_context.return_value = ssl_mock
    with mocker.patch("builtins.open", mocker.mock_open()) as mock_file:
        assert load_config() == 1
        mock_file.assert_called_once_with(
            "/.cache/pre-commit/oaf_pre-commit_config.json", "w"
        )
        mock_file().write.assert_called_once_with(urlopen_mock.read())


# Tests that the function loads the configuration file from the cache directory when the live configuration is empty. tags: [happy path]
def test_config_load_from_cache(self, monkeypatch, tmp_path):
    oaf_config = {"cache": [], "live": []}
    monkeypatch.delenv("PRE_COMMIT_HOME", raising=False)
    monkeypatch.delenv("XDG_CACHE_HOME", raising=False)
    oaf_config["cache"] = {
        "OAF_GIT_BRANCH_NAME_REGEX": "^((release\/[0-9]+\.[0-9]+\.[0-9])|((feature|feat|cleanup|bugfix|hotfix|fix|devops)\/[A-Z]+-[0-9]+(-[A-z0-9]+)+)|(umuganda|chore)\/[a-zA-Z0-9+\-]+)$",
        "OAF_WATCH_COMMIT_HISTORY": False,
        "OAF_GIT_BRANCH_NAME_EXCEPTION": ["develop", "main", "master"],
        "OAF_GIT_COMMIT_TYPES": [
            "feat",
            "fix",
            "style",
            "refactor",
            "docs",
            "perf",
            "test",
            "chore",
            "ci",
            "build",
        ],
        "OAF_REQUIRED_HOOKS": {
            "ggshield": {
                "args": ["--help"],
                "repo": "https://github.com/gitguardian/ggshield",
            },
            "gitlint": {
                "args": ["--help"],
                "repo": "https://github.com/jorisroovers/gitlint",
            },
        },
    }
    oaf_config["live"] = []
    config_file_path = tmp_path / ".cache/pre-commit/oaf_pre-commit_config.json"
    with open(config_file_path, "w") as f:
        json.dump(oaf_config["cache"], f)
    assert load_config() == 5


# Tests that the function constructs the correct command string when given arguments. tags: [happy path]
def test_hook_installed_with_arguments(self):
    assert is_hook_installed("pre-commit", ["--all-files"]) == True
    assert is_hook_installed("flake8", ["--max-line-length=80"]) == True


# Tests that the function constructs the correct command string when given no arguments. tags: [happy path]
def test_hook_installed_without_arguments(self):
    assert is_hook_installed("black", []) == True
    assert is_hook_installed("isort", []) == True


# Tests that the function returns false when the hook is not installed. tags: [edge case]
def test_hook_installed_not_installed(self):
    assert is_hook_installed("non-existent-hook", []) == False


# Tests that the function returns false when the hook name is empty. tags: [edge case]
def test_hook_installed_empty_name(self):
    assert is_hook_installed("", []) == False


# Tests that the function handles special characters in hook name and arguments correctly. tags: [general behavior]
def test_hook_installed_special_characters(self):
    assert is_hook_installed("pre-commit", ["--hook-type=python"]) == True
    assert is_hook_installed("pytest", ["-k", "test_functional"]) == True


# Tests that the function handles spaces in hook name and arguments correctly. tags: [general behavior]
def test_hook_installed_spaces(self):
    assert (
        is_hook_installed("pre-commit", ["--hook-type=python", "--exclude=tests"])
        == True
    )
    assert is_hook_installed("pytest", ["-m", "not slow"]) == True


def test_get_current_branch_name_on_branch(self, mocker):
    mocker.patch("subprocess.getoutput", return_value="main")
    assert get_current_branch_name() == "main"


# Tests that the function returns an empty string when the git repository is empty. tags: [edge case]
def test_get_current_branch_name_empty_repository(self, mocker):
    mocker.patch(
        "subprocess.getoutput",
        return_value="fatal: not a git repository (or any of the parent directories): .git",
    )
    assert get_current_branch_name() == ""


# Tests that the function returns an empty string when the git repository has no branches. tags: [edge case]
def test_get_current_branch_name_no_branches(self, mocker):
    mocker.patch("subprocess.getoutput", return_value="HEAD")
    assert get_current_branch_name() == ""


# Tests that the function returns "head" when the git repository is in a detached head state. tags: [edge case]
def test_get_current_branch_name_detached_head(self, mocker):
    mocker.patch("subprocess.getoutput", return_value="HEAD")
    assert get_current_branch_name() == "HEAD"


# Tests that the function raises a subprocess.calledprocesserror when git is not installed on the system. tags: [edge case]
def test_get_current_branch_name_git_not_installed(self, mocker):
    mocker.patch(
        "subprocess.getoutput", side_effect=subprocess.CalledProcessError(1, "git")
    )
    with pytest.raises(subprocess.CalledProcessError):
        get_current_branch_name()


# Tests that the function raises a subprocess.calledprocesserror when the user running the script does not have permission to access the git repository. tags: [edge case]
def test_get_current_branch_name_permission_denied(self, mocker):
    mocker.patch(
        "subprocess.getoutput",
        side_effect=subprocess.CalledProcessError(1, "git rev-parse --abbrev-ref HEAD"),
    )
    with pytest.raises(subprocess.CalledProcessError):
        get_current_branch_name()


# Tests that a commit title with a valid conventional commit type is correctly validated. tags: [happy path]
def test_valid_commit_title(self):
    commit = {"title": "feat: add new feature", "hash": "abc123"}
    assert validate_git_commit(commit) == True


# Tests that a commit title with a valid conventional commit type and parentheses is correctly validated. tags: [happy path]
def test_valid_commit_title_with_parentheses(self):
    commit = {"title": "fix(docs): update documentation", "hash": "def456"}
    assert validate_git_commit(commit) == True


# Tests that a commit title with an invalid conventional commit type is correctly identified as invalid. tags: [edge case]
def test_commit_title_with_invalid_conventional_commit_type(self):
    commit = {"title": "invalid: this is not a valid commit type", "hash": "ghi789"}
    assert validate_git_commit(commit) == False


# Tests that an empty oaf_config["cache"] dictionary is handled correctly. tags: [edge case]
def test_empty_oaf_config_cache(self, monkeypatch):
    oaf_config = {"cache": []}
    monkeypatch.setitem(oaf_config, "cache", {})
    commit = {"title": "feat: add new feature", "hash": "abc123"}
    assert validate_git_commit(commit) == False


# Tests that a commit title with no separator is correctly identified as invalid. tags: [edge case]
def test_commit_title_with_no_separator(self):
    commit = {"title": "this is not a valid commit title", "hash": "jkl012"}
    assert validate_git_commit(commit) == False


# Tests that a commit title with an empty first part is correctly identified as invalid. tags: [edge case]
def test_commit_title_with_empty_first_part(self):
    commit = {"title": ": this is not a valid commit title", "hash": "mno345"}
    assert validate_git_commit(commit) == False


# Tests that the function correctly extracts and formats the commit title and message when the git repository has only one commit. tags: [happy path]
def test_get_commits_single_commit(self, mocker):
    mocker.patch(
        "subprocess.check_output",
        return_value=b"commit abc123\nAuthor: John Doe\nDate: 2021-01-01\n\nThis is a commit message.",
    )
    commits = get_commits()
    assert len(commits) == 1
    assert commits[0]["hash"] == "abc123"
    assert commits[0]["title"] == "This is a commit message."
    assert commits[0]["message"] == ""


# Tests that the function correctly handles multi-line commit messages. tags: [happy path]
def test_get_commits_multi_line_commit_messages(self, mocker):
    mocker.patch(
        "subprocess.check_output",
        return_value=b"commit abc123\nAuthor: John Doe\nDate: 2021-01-01\n\nThis is a commit message.\n\nThis is a second line.",
    )
    commits = get_commits()
    assert len(commits) == 1
    assert commits[0]["hash"] == "abc123"
    assert commits[0]["title"] == "This is a commit message."
    assert commits[0]["message"] == "This is a second line."


# Tests that the function returns an empty list when the git repository has no commits. tags: [edge case]
def test_get_commits_empty_repository(self, mocker):
    mocker.patch("subprocess.check_output", return_value=b"")
    commits = get_commits()
    assert len(commits) == 0


# Tests that the function does not work correctly when the git repository is not in a valid state. tags: [edge case]
def test_get_commits_invalid_git_repository(self, mocker):
    mocker.patch(
        "subprocess.check_output",
        side_effect=subprocess.CalledProcessError(1, "git log"),
    )
    with pytest.raises(subprocess.CalledProcessError):
        get_commits()


# Tests that the function correctly handles commits with non-standard characters in commit message. tags: [edge case]
def test_get_commits_non_standard_characters_in_commit_message(self, mocker):
    mocker.patch(
        "subprocess.check_output",
        return_value="commit abc123\nAuthor: John Doe\nDate: 2021-01-01\n\nThis is a commit message with non-standard characters: é, ü, ñ.",
    )
    commits = get_commits()
    assert len(commits) == 1
    assert commits[0]["hash"] == "abc123"
    assert (
        commits[0]["title"]
        == "This is a commit message with non-standard characters: é, ü, ñ."
    )
    assert commits[0]["message"] == ""


# Tests that the function correctly handles commits with non-standard characters in author name. tags: [edge case]
def test_get_commits_non_standard_characters_in_author_name(self, mocker):
    mocker.patch(
        "subprocess.check_output",
        return_value="commit abc123\nAuthor: Jöhn Döe\nDate: 2021-01-01\n\nThis is a commit message.",
    )
    commits = get_commits()
    assert len(commits) == 1
    assert commits[0]["hash"] == "abc123"
    assert commits[0]["title"] == "This is a commit message."
    assert commits[0]["message"] == ""


tests/oaf_pre_commit_hook_git_branch_test.py
from __future__ import annotations

import shutil

import pytest


pre_commit_hooks/utils.py
import json
import os
import ssl
import subprocess
import warnings
from urllib.request import urlopen


def load_config() -> dict:
    CONFIG_URL = "https://raw.githubusercontent.com/one-acre-fund/oaf-pre-commit-hooks/main/config.json"
    pre_commit_home = os.getenv("PRE_COMMIT_HOME")
    cache_dir = ""
    if pre_commit_home is None or os.path.exists(pre_commit_home) == False:
        pre_commit_home = (
            "" if os.getenv("XDG_CACHE_HOME") is None else os.getenv("XDG_CACHE_HOME")
        )
        cache_dir = pre_commit_home + "/pre-commit"
        if os.path.exists(cache_dir) == False:
            user_home = os.path.expanduser("~")
            cache_dir = user_home + "/.cache/pre-commit"
            if os.path.exists(cache_dir) == False:
                os.makedirs(cache_dir)

    config_file_path = cache_dir + "/oaf_pre-commit_config.json"
    config = {}

    if os.path.exists(config_file_path):
        with open(config_file_path) as json_file:
            config = json.load(json_file)
    if len(config) < 1:
        try:
            ssl._create_default_https_context = ssl._create_unverified_context
            with urlopen(CONFIG_URL) as f:
                config = json.load(f)
                print(" from %s %o", CONFIG_URL, config)
                if len(config) > 1:
                    with open(config_file_path, "w") as fp:
                        json.dump(config, fp)
        except Exception as e:
            warnings.warn(e)

    if len(config) < 1:
        config = {
            "OAF_GIT_BRANCH_NAME_REGEX": "^((release\/[0-9]+\.[0-9]+\.[0-9])|((feature|feat|cleanup|bugfix|hotfix|fix|devops)\/[A-Z]+-[0-9]+(-[A-z0-9]+)+)|(umuganda|chore)\/[a-zA-Z0-9+\-]+)$",
            "OAF_GIT_BRANCH_NAME_EXCEPTION": ["develop", "main", "master"],
            "OAF_GIT_COMMIT_TYPES": [
                "feat",
                "fix",
                "style",
                "refactor",
                "docs",
                "perf",
                "test",
                "chore",
                "build",
            ],
        }
    return config


def is_hook_installed(hook) -> bool:
    """Determine if the pre-commit hook is installed and enabled."""
    exit_code = os.WEXITSTATUS(os.system("git config --bool " + hook))
    out = subprocess.getoutput("git config --bool " + hook)
    return exit_code == 0 and out == "false"


pre_commit_hooks/oaf_pre_commit_hook_git_commit.py
#!/usr/bin/env python3
"""Helper script to be used as a pre-commit hook."""
import argparse
import os
import re
import ssl
import subprocess
import sys
import warnings
from typing import Sequence
from urllib.request import urlopen

TERMINAL_COLOR_ERROR = "\033[1;31;40m"
TERMINAL_COLOR_WARNING = "\033[1;33;40m"
TERMINAL_COLOR_NORMAL = "\033[0;37;40m"
TERMINAL_COLOR_PASS = "\033[1;32;40m"
oaf_config = {"cache": [], "live": []}


def load_config() -> int:
    pre_commit_home = os.getenv("PRE_COMMIT_HOME")
    cache_dir = ""
    if pre_commit_home is None or os.path.exists(pre_commit_home) == False:
        pre_commit_home = (
            "" if os.getenv("XDG_CACHE_HOME") is None else os.getenv("XDG_CACHE_HOME")
        )
        cache_dir = pre_commit_home + "/pre-commit"
        if os.path.exists(cache_dir) == False:
            user_home = os.path.expanduser("~")
            cache_dir = user_home + "/.cache/pre-commit"
            if os.path.exists(cache_dir) == False:
                os.makedirs(cache_dir)

    config_file_path = cache_dir + "/oaf_pre-commit_config.json"

    try:
        config_url = "https://raw.githubusercontent.com/one-acre-fund/oaf-pre-commit-hooks/main/config.json"
        ssl._create_default_https_context = ssl._create_unverified_context
        with urlopen(config_url) as f:
            oaf_config["live"] = json.load(f)
            if len(oaf_config["live"]) > 1:
                oaf_config["cache"] = oaf_config["live"]
                with open(config_file_path, "w") as fp:
                    json.dump(oaf_config["live"], fp)
    except Exception as e:
        print(
            "%sFailed to get config from %s %s"
            % (TERMINAL_COLOR_ERROR, config_url, TERMINAL_COLOR_NORMAL)
        )
        print("%s trace: %s %s" % (TERMINAL_COLOR_WARNING, e, TERMINAL_COLOR_NORMAL))

    if len(oaf_config["live"]) < 1 and os.path.exists(config_file_path):
        with open(config_file_path) as json_file:
            oaf_config["cache"] = json.load(json_file)

    if oaf_config["cache"] and len(oaf_config["live"]) < 1:
        oaf_config["cache"] = oaf_config["live"] = {
            "OAF_GIT_BRANCH_NAME_REGEX": "^((release\/[0-9]+\.[0-9]+\.[0-9])|((feature|feat|cleanup|bugfix|hotfix|fix|devops)\/[A-Z]+-[0-9]+(-[A-z0-9]+)+)|(umuganda|chore)\/[a-zA-Z0-9+\-]+)$",
            "OAF_GIT_BRANCH_NAME_EXCEPTION": ["develop", "main", "master"],
            "OAF_GIT_COMMIT_TYPES": [
                "feat",
                "fix",
                "style",
                "refactor",
                "docs",
                "perf",
                "test",
                "chore",
                "build",
            ],
            "OAF_WATCH_COMMIT_HISTORY": False,
        }
    print(oaf_config)
    return len(oaf_config)


def is_hook_installed(hook) -> bool:
    """Determine if the pre-commit hook is installed and enabled."""
    exit_code = os.WEXITSTATUS(os.system("git config --bool " + hook))
    out = subprocess.getoutput("git config --bool " + hook)
    return exit_code == 0 and out == "false"


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--config", default="", help="check README.md")
    parser.add_argument("--forced", default=True, help="check README.md")
    parser.add_argument("--verbose", default=True, help="Verbose debug logging.")
    args, unknown_args = parser.parse_known_args(argv)

    if args.verbose:
        print(
            "%s %2d files to check %s"
            % (TERMINAL_COLOR_PASS, len(unknown_args), TERMINAL_COLOR_NORMAL)
        )
        print(args, unknown_args)

    if len(sys.argv) == 1:
        print("How to pass args when calling the script:")
        parser.print_help()

    if (
        args.forced == False
        and is_hook_installed("hooks.oaf-pre-commit-hook-git-commit") == False
    ):
        print(
            "%sHook is not installed or disabled:(enable with 'git config hooks.oaf-pre-commit-hook-git-branch true') %s"
            % (TERMINAL_COLOR_ERROR, TERMINAL_COLOR_NORMAL)
        )
        return 1

    return 0


if __name__ == "__main__":
    raise SystemExit(main())


pre_commit_hooks/oaf_tech_pre_commit_hook.py
#!/usr/bin/env python3
"""Helper script to be used as a pre-commit hook."""
import argparse
import json
from ruamel import yaml
import os
import re
import ssl
import subprocess
import sys
from typing import Sequence
from urllib.request import urlopen

TERMINAL_COLOR_ERROR = "\033[1;31;40m"
TERMINAL_COLOR_WARNING = "\033[1;33;40m"
TERMINAL_COLOR_NORMAL = "\033[0;37;40m"
TERMINAL_COLOR_PASS = "\033[1;32;40m"
oaf_config = {"cache": {}, "live": {}}


def load_config(use_cache=True) -> int:
    pre_commit_home = os.getenv("PRE_COMMIT_HOME")
    if pre_commit_home is None or os.path.exists(pre_commit_home) == False:
        if (
            os.getenv("XDG_CACHE_HOME") is not None
            and len(os.getenv("XDG_CACHE_HOME")) >= 1
        ):
            pre_commit_home = os.getenv("XDG_CACHE_HOME") + "/pre-commit"

        if pre_commit_home is None or os.path.exists(pre_commit_home) == False:
            user_home = os.path.expanduser("~")
            pre_commit_home = user_home + "/.cache/pre-commit"
            if os.path.exists(pre_commit_home) == False:
                os.makedirs(pre_commit_home)

    config_file_path = pre_commit_home + "/oaf_pre-commit_config.json"
    try:
        config_url = "https://raw.githubusercontent.com/one-acre-fund/oaf-pre-commit-hooks/main/config.json"
        ssl._create_default_https_context = ssl._create_unverified_context
        with urlopen(config_url) as f:
            oaf_config["live"] = json.load(f)
            if len(oaf_config["live"]) > 1:
                oaf_config["cache"] = oaf_config["live"]
                with open(config_file_path, "w") as fp:
                    json.dump(oaf_config["live"], fp)
    except Exception as e:
        print(
            "%sFailed to get config from %s while cache= %s %s"
            % (TERMINAL_COLOR_ERROR, config_url, use_cache, TERMINAL_COLOR_NORMAL)
        )
        print("%s trace: %s %s" % (TERMINAL_COLOR_WARNING, e, TERMINAL_COLOR_NORMAL))

    if len(oaf_config["live"]) < 1 and os.path.exists(config_file_path):
        with open(config_file_path) as json_file:
            oaf_config["cache"] = json.load(json_file)

    if len(oaf_config["cache"]) < 1 and len(oaf_config["live"]) < 1:
        oaf_config["cache"] = oaf_config["live"] = {
            "OAF_GIT_BRANCH_NAME_REGEX": "^((release\/[0-9]+\.[0-9]+\.[0-9])|((feature|feat|cleanup|bugfix|hotfix|fix|devops)\/[A-Z]+-[0-9]+(-[A-z0-9]+)+)|(umuganda|chore)\/[a-zA-Z0-9+\-]+)$",
            "OAF_WATCH_COMMIT_HISTORY": False,
            "OAF_GIT_BRANCH_NAME_EXCEPTION": ["develop", "main", "master"],
            "OAF_GIT_COMMIT_TYPES": [
                "feat",
                "fix",
                "style",
                "refactor",
                "docs",
                "perf",
                "test",
                "chore",
                "ci",
                "build",
            ],
            "OAF_REQUIRED_HOOKS": {
                "ggshield": {
                    "args": ["--verbose"],
                    "repo": "https://github.com/gitguardian/ggshield",
                },
                "gitlint": {
                    "args": ["--verbose"],
                    "repo": "https://github.com/jorisroovers/gitlint",
                },
                "markdownlint": {
                    "args": [
                        "--verbose",
                    ],
                    "repo": "https://github.com/jorisroovers/gitlint",
                },
            },
        }
    return len(oaf_config)


def contains_config_directive(cfg, dir):
    return re.findall(dir, cfg, re.M)


def is_hook_installed_cli(hook, info) -> bool:
    """Determine if the pre-commit hook is installed and enabled using CLI"""
    arg_str = " ".join(info["args"])
    cmd = "pre-commit run " + hook + " " + arg_str.strip() + ""
    exit_code = os.WEXITSTATUS(os.system(cmd))
    exit_output = subprocess.getoutput(cmd)
    no_hook_found = exit_output.find("No hook with id")
    print(cmd, no_hook_found, exit_code)
    return no_hook_found == -1


def is_hook_installed_config(hook, info) -> bool:
    """Determine if the pre-commit hook is installed and enabled by config YAML"""
    try:
        with open(".pre-commit-config.yaml", "r") as stream:
            config = yaml.safe_load(stream)
            if config["repos"] is not None:
                for repo in config["repos"]:
                    for config_hook in repo["hooks"]:
                        if hook == config_hook["id"]:
                            return repo["repo"] == info["repo"]
    except Exception as e:
        print(e)
    return False


def get_current_branch_name() -> str:
    """Determine the active Git branch name."""
    return subprocess.getoutput("git rev-parse --abbrev-ref HEAD")


def get_git_branch_name_regex() -> str:
    """Read Git and GitFlow naming Regex"""
    if len(oaf_config["cache"]) < 1:
        load_config()
    return oaf_config["cache"]["OAF_GIT_BRANCH_NAME_REGEX"]


def get_git_branch_name_exceptions() -> list:
    """Determine Git branches that are exempt from naming convention"""

    if len(oaf_config["cache"]) < 1:
        load_config()

    return oaf_config["cache"]["OAF_GIT_BRANCH_NAME_EXCEPTION"]


def get_git_conventional_commit_types() -> list:
    """Determine Git conventional commit types"""
    if len(oaf_config["cache"]) < 1:
        load_config()

    return oaf_config["cache"]["OAF_GIT_COMMIT_TYPES"]


def validate_git_commit(commit, verbose=False) -> bool:
    """Parse and verify git commit format"""
    commit_title = commit["title"].split(":")
    if len(commit_title) == 1:
        if verbose:
            print(
                "%sCommit %s has invalid message `%s` %s"
                % (
                    TERMINAL_COLOR_WARNING,
                    commit["hash"],
                    commit["title"],
                    TERMINAL_COLOR_NORMAL,
                )
            )
        return False

    is_commit_ok = False
    oaf_commit_types = get_git_conventional_commit_types()
    commit_type = commit_title[0].split("(")[0]
    try:
        is_commit_ok = oaf_commit_types.index(commit_type) >= 0
    except ValueError:
        if verbose:
            print(
                "%sCommit %s '%s' has wrong type `%s` %s"
                % (
                    TERMINAL_COLOR_WARNING,
                    commit["hash"],
                    commit["title"],
                    commit_type,
                    TERMINAL_COLOR_NORMAL,
                )
            )
        return False

    return is_commit_ok


def get_commits() -> list:
    """Get commit history on current branch"""
    lines = (
        subprocess.check_output(["git", "log"], stderr=subprocess.STDOUT)
        .decode("utf-8")
        .split("\n")
    )
    commits = []
    current_commit = {}

    def save_current_commit():
        title = current_commit["message"][0]
        message = current_commit["message"][1:]
        if message and message[0] == "":
            del message[0]
        current_commit["title"] = title.lstrip()
        current_commit["message"] = "\n".join(message)
        commits.append(current_commit)

    for line in lines:
        if not line.startswith(" "):
            if line.startswith("commit "):
                if current_commit:
                    save_current_commit()
                    current_commit = {}
                current_commit["hash"] = line.split("commit ")[1]
            else:
                try:
                    key, value = line.split(":", 1)
                    current_commit[key.lower()] = value.strip()
                except ValueError:
                    pass
        else:
            current_commit.setdefault("message", []).append(
                re.compile("^`{4}`").sub("", line)
            )
    if current_commit:
        save_current_commit()
    return commits


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--forced", default="False", help="check README.md")

    args, unknown_args = parser.parse_known_args(argv)
    print(
        "%s %2d files to check %s"
        % (TERMINAL_COLOR_PASS, len(unknown_args), TERMINAL_COLOR_NORMAL)
    )
    if args.forced == True:
        print("Forced mode: ")

    # check on required `pre-commit` hooks
    if len(oaf_config["cache"]) < 1:
        load_config()

    for hook in oaf_config["cache"]["OAF_REQUIRED_HOOKS"]:
        hook_info = oaf_config["cache"]["OAF_REQUIRED_HOOKS"][hook]
        if hook_info is None or is_hook_installed_config(hook, hook_info) == False:
            print(
                "%s hook %s is not installed or is disabled %s see more:%s"
                % (
                    TERMINAL_COLOR_ERROR,
                    hook,
                    hook_info["repo"],
                    TERMINAL_COLOR_NORMAL,
                )
            )
            return 1

    # validate branch naming
    if len(oaf_config["cache"]) < 1:
        load_config()

    branch = get_current_branch_name()
    oaf_lts_branches = get_git_branch_name_exceptions()
    try:
        is_branch_ok = is_branch_lts = oaf_lts_branches.index(branch) >= 0
    except ValueError:
        is_branch_ok = is_branch_lts = False

    if is_branch_lts == False:
        oaf_regex = get_git_branch_name_regex()
        is_branch_ok = re.search(oaf_regex, branch)

    if is_branch_ok is None or is_branch_ok == False:
        print(
            "%sBranch `%s` should follow name={prefix/JIRA#-descr}:"
            "prefix=%s or name=%s %s"
            % (
                TERMINAL_COLOR_ERROR,
                branch,
                oaf_config["cache"]["OAF_GIT_BRANCH_NAME_REGEX"],
                ",".join(oaf_lts_branches),
                TERMINAL_COLOR_NORMAL,
            )
        )
        return 1

    # check commit history on this branch
    if len(oaf_config["cache"]) < 1:
        load_config()

    if oaf_config["cache"]["OAF_WATCH_COMMIT_HISTORY"]:
        commits = get_commits()
        for commit in commits:
            is_commit_ok = validate_git_commit(commit)
            if is_commit_ok == False and is_branch_lts == False:
                return 3

    # check current commit message (e.g. prepare-commit-msg) using gitlint
    if len(oaf_config["cache"]) < 1:
        load_config()

    hook_info = oaf_config["cache"]["OAF_REQUIRED_HOOKS"]["gitlint"]
    if hook_info is None or is_hook_installed_config("gitlint", hook_info) == False:
        return 4
    else:
        # check on .gitlint
        gitlint_path = subprocess.getoutput("git rev-parse --show-toplevel")
        gitlint_path += "/.gitlint"
        try:
            if os.path.exists(gitlint_path) == False:
                print(
                    "%s installing .gitlint at %s %s"
                    % (TERMINAL_COLOR_WARNING, gitlint_path, TERMINAL_COLOR_NORMAL)
                )
                gitlint_url = "https://raw.githubusercontent.com/one-acre-fund/oaf-pre-commit-hooks/main/.gitlint"
                ssl._create_default_https_context = ssl._create_unverified_context
                with urlopen(gitlint_url) as gitlint_file:
                    gitlint_config = str(gitlint_file.read(), "UTF-8")
                    f = open(gitlint_path, "w")
                    f.write(gitlint_config)
                    f.close()
            else:
                keywords = [
                    "contrib=contrib-title-conventional-commits",
                    "types=",
                    "title-min-length",
                    "title-max-length",
                    "body-max-line-length",
                ]
                gitlint_file = open(gitlint_path, "r")
                gitlint_config = gitlint_file.read()
                for keyword in keywords:
                    matching_regex = contains_config_directive(gitlint_config, keyword)
                    if len(matching_regex) < 1:
                        print(
                            "%s .gitlint is missing important rules %s %s %s"
                            % (
                                TERMINAL_COLOR_ERROR,
                                gitlint_path,
                                gitlint_config,
                                TERMINAL_COLOR_NORMAL,
                            )
                        )
                        gitlint_file.close()
                        return 4
        except Exception as e:
            print(
                "%sFailed to read .gitlint config from %s : %s %s"
                % (TERMINAL_COLOR_ERROR, gitlint_path, e, TERMINAL_COLOR_NORMAL)
            )
            return 4
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
